content/shmem_test_some.tex

```diff
@@ -0,0 +1,77 @@
+\apisummary{
+  Test whether some number of variables on the local \ac{PE} have changed to a
+  specified value.
+}
+
+\begin{apidefinition}
+
+\begin{C11synopsis}
+size_t @\FuncDecl{shmem\_test\_some}@(TYPE *ivars, size_t nelems, _Bool *status, int cmp, TYPE cmp_value);
+\end{C11synopsis}
+where \TYPE{} is one of the point-to-point synchronization types specified by
+Table \ref{p2psynctypes}.
+
+\begin{Csynopsis}
+size_t @\FuncDecl{shmem\_\FuncParam{TYPENAME}\_test\_some}@(TYPE *ivars, size_t nelems, _Bool *status, int cmp,
+    TYPE cmp_value);
+\end{Csynopsis}
+where \TYPE{} is one of the point-to-point synchronization types and has a
+corresponding \TYPENAME{} specified by Table \ref{p2psynctypes}.
+
+\begin{apiarguments}
+
+  \apiargument{OUT}{ivars}{A pointer to an array of remotely accessible data
+    objects. The type of \VAR{ivars} should match that implied in the SYNOPSIS
+    section.} 
+  \apiargument{IN}{nelems}{The number of elements in the \VAR{ivars} array to be
+    compared with \VAR{cmp\_value}.}
+  \apiargument{INOUT}{status}{A mask array of length \VAR{nelems}, where each
+    element corresponds to the respective element in \VAR{ivars} and represents
+    the status of each implied wait condition.  On input, elements of \VAR{status}
+    set to 0 will be tested, and elements set to 1 will be ignored.  On output,
+    each element of \VAR{status} equal to 1 corresponds to a satisfied
+    condition in the respective element of \VAR{ivars} (or an ignored
+    condition, if initialized to 1 on input).}
```

Comment:
Yup, this is exactly the problem.  IIRC, some in the WG greatly preferred that `status` have a boolean type rather than some stranger alternative which supports the 3 states.  I do believe that committing to 0/1 `status` elements leads to two options: 1) adjust the `status` array each iteration or 2) keep track of a separate "completed" array.  I agree that neither is "the correct way", but I settled on #2 because I like the convenience of being able to reuse the `status` array, and I kinda think it's a little easier to follow.  But, yes, it's a bit of a wash...  I also agree with @jamesaross that the code example looks like bad practice at a glance.  Do you think Option #1 could improve that?

I'm thinking of re-writing the proposal with Option 1 to see how it looks.  Does that sound good?

Reply:
I apologize for not taking a closer look at the proposal earlier in the process. I think that I expected a prototype (and behavioral function) more like:

```c
size_t shmem_test_some(TYPE *ivars, size_t nelems, size_t *satisfied,
                       int cmp, TYPE cmp_value)
{
  size_t count = 0;
  for (size_t i = 0; i < nelems; i++) {
    switch (cmp) {
    case SHMEM_CMP_EQ:
      if (ivars[i] == cmp_value)
        satisfied[count++] = i;
      break;
    /* case ... */
    }
  }
  return count;
}
```

This way, the call to `shmem_test_some` requires _O_(nelems) accesses to check the array of values, but the user of the call only needs _O_(satisfied) accesses.

Is this "better" than the current proposal (under either Option 1 or Option 2)? In some cases, I think it is because there are fewer memory references in user code. What this alternate interface lacks is the "mask" feature of the current proposal. To resolve this, a mask array could be provided as an extra argument to the function with a null-check fast path when the user wants to unconditionally check the entire array.

Reply:
Yeah, this is actually quite similar to what was presented in my first set of WG slides [here](https://docs.google.com/presentation/d/1vp1XWFE0E2JMUDA5lEdXngEhazETnhJRPwpwDmcqYTM/edit?usp=sharing), but somewhere along the line the `indices`/`satisfied` array morphed into the masked version, so that we could feed it back to the routine to better support task-processing-like patterns.  But yes, this is probably better, especially with the optional mask array idea, which I like.